contracts/pallet-ibc/src/ics20_fee/mod.rs

```diff
@@ -0,0 +1,281 @@
+// Copyright 2022 ComposableFi
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::routing::Context;
+use alloc::{format, string::ToString};
+use core::fmt::Formatter;
+use ibc::{
+	applications::transfer::{
+		context::BankKeeper, is_receiver_chain_source, packet::PacketData, TracePrefix,
+	},
+	core::{
+		ics04_channel::{
+			channel::{Counterparty, Order},
+			error::Error as Ics04Error,
+			msgs::acknowledgement::Acknowledgement,
+			packet::Packet,
+			Version,
+		},
+		ics24_host::identifier::{ChannelId, ConnectionId, PortId},
+		ics26_routing::context::{Module, ModuleCallbackContext, ModuleOutputBuilder},
+	},
+	signer::Signer,
+};
+use ibc_primitives::IbcAccount;
+use sp_core::crypto::AccountId32;
+use sp_runtime::traits::{AccountIdConversion, Get};
+use sp_std::marker::PhantomData;
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use frame_support::{pallet_prelude::*, PalletId};
+	use frame_system::pallet_prelude::OriginFor;
+	use sp_runtime::{traits::Get, Perbill};
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + crate::Config {
+		#[pallet::constant]
+		type ServiceCharge: Get<Perbill>;
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub (super) trait Store)]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	#[allow(clippy::disallowed_types)]
+	pub type ServiceCharge<T: Config> = StorageValue<_, Perbill, OptionQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(0)]
+		pub fn set_charge(origin: OriginFor<T>, charge: u32) -> DispatchResult {
+			<T as crate::Config>::AdminOrigin::ensure_origin(origin)?;
+			<ServiceCharge<T>>::put(Perbill::from_percent(charge));
```

Comment:
can we add a validation here that at least is less than 1% for that charge in percent?

Reply:
I think just for code hygiene, a magic max constant isn't necessary

Reply:
May be a runtime configuration may help, if the maximum charge rate is of concern